### ISSUE TITLE
IVRDriverInput.Update*Component's fTimeOffset is ignored?

### ISSUE THREAD

**Issue by bddckr** [ValveSoftware/openvr#760]
I'm currently only using `IVRDriverInput.UpdateBooleanComponent` and have to basically set it to `true` and `false` right after (hardware doesn't give me any up+down, only a "clicked").

Using `fTimeOffset`s set to `0.0` and `0.1` respectively doesn't seem to make the runtime notice the `/click` of that boolean component.

If I put this logic into a new `std::thread` and wait via `std::this_thread::sleep_for(std::chrono::milliseconds(100));` between the two `UpdateBooleanComponent` calls it works just fine, though.

Is there anything special to do here or is it a bug? I tried running the update logic inside `IServerTrackedDeviceProvider.RunFrame` and `ITrackedDeviceServerDriver.DebugRequest` with the same issue above.

**Comment by peroht** [comment]
Not sure I understand fully if you are using a custom hardware or not - but from what I've experienced with the normal hand controllers it is that if you want to send a /click you need to send a /touch first. I've only used fTimeOffset to 0.0.

**Comment by bddckr** [comment]
@peroht Are you able to do this, or are you setting true and false at different times in response to your hardware giving you a down/up event? On my hardware I don't get down+up, I only get a click, so that's why I'm trying to let OpenVR know about the state changing to true and false right after, at the same time.
```c++
pDriverInput->UpdateBooleanComponent(m_ulInputComponentHandle, true, 0);
pDriverInput->UpdateBooleanComponent(m_ulInputComponentHandle, false, 0);
```

I'll create another boolean component for `/touch` and will try to send that first, thanks for that. I'm expecting it to not help, though, because the code above works if I put in a thread sleep - I'd expect that to not work at all until the `/touch` is added if that's the case...

**Comment by JoeLudwig** [comment]
What is your component called?

At the moment the binding files that map the component names from the driver into something that the application can see are somewhat limited. There's also no UI to edit them yet. So unless you use one of the names listed here there's probably no one listening: https://github.com/ValveSoftware/openvr/wiki/IVRDriverInput-Overview

Both of those limitations should be corrected soon.

**Comment by bddckr** [comment]
I'm using `/input/system/click` and additionally have set the input profile path to a `.json` file:
```js
{
	"jsonid": "input_profile",
	"controller_type": "vive_controller",
	"legacy_profile": "vive_controller"
}
```

Again - this works fine if I have a thread running and do the above with a delay in between:
```c++
pDriverInput->UpdateBooleanComponent(m_ulInputComponentHandle, true, 0);
std::this_thread::sleep_for(std::chrono::milliseconds(100));
pDriverInput->UpdateBooleanComponent(m_ulInputComponentHandle, false, 0);
```
Which is why I don't think the issue lies elsewhere.

The old pre-`IVRDriverInput` API also doesn't work without a thread sleep in between the calls (which makes sense as the old API is routed through new input handling system):
```c++
VRServerDriverHost()->TrackedDeviceButtonPressed(0, k_EButton_System, 0);
VRServerDriverHost()->TrackedDeviceButtonUnpressed(0, k_EButton_System, 0.1);
```

**Comment by peroht** [comment]
@bddckr I am not using the /click so I cannot comment on that. Also, I am not sending a true followed by an immediate false I have other controller component updates in between the touch/untouch and press/unpress with slight delays inbetween (id expect something between 10 to 50 ms inbetween the change of states). I hope youll get it sorted.

**Comment by peroht** [comment]
@JoeLudwig looking forward to an initial UI for this. Also I hope that more games/engines will honor the `TrackedControllerRole_OptOut ` soon.

**Comment by bddckr** [comment]
@peroht Thanks for that!

I can't seem to get this to work without the artificial delay (thread sleep). I thought the `fTimeOffset` is usable here but it doesn't work.

I really think this is a regression in the runtime as previously just doing `TrackedDeviceButtonPressed` with the time offsets worked. This was back in December.

I'll go with the thread sleep for now. BTW: A delay of 50ms also works, but only sporadically. 100ms always works, so I'm going with that for now.

---

`TrackedControllerRole_OptOut` and `Prop_NeverTracked_Bool` are nice properties. I also hope we see updates to the UE4 and Unity plugins soon so more game devs support it in their games!

**Comment by JoeLudwig** [comment]
Any app using controller roles would need to call GetTrackedDeviceIndexForControllerRole and/or GetControllerRoleForTrackedDeviceIndex to do that. Those functions should already be obeying TrackedControllerRole_OptOut. Are you seeing a case where they aren't?

We added Prop_NeverTracked_Bool so that the status window would stop blinking annoyingly for gamepads. What support for that property would you expect in applications?

The timestamp in UpdateBooleanComponent is generally meant to be negative to mean the state change happened some time in the past. It's there to allow drivers to report how much USB or wireless latency they're dealing with so that poses can be predicted back (interpolated?) to the appropriate time to match user events for throwing and shooting. As you've discovered, times in the future don't delay the arrival of the new information, they just set the internal timestamp forward.

That you're calling UpdateBooleanComponent with true, then false makes me think you're not actually reflecting button input with that component. What kind of physical hardware ARE you reflecting?

**Comment by bddckr** [comment]
> That you're calling UpdateBooleanComponent with true, then false makes me think you're not actually reflecting button input with that component. What kind of physical hardware ARE you reflecting?

I don't have a lot of details I can share, it's mostly that the API I'm using for the hardware is only telling me about when the button was **unpressed**. I.e. there is no "down" state.

Negative timestamps make sense! I just checked and the following doesn't work, still needs a thread sleep in between the calls:
```c++
pDriverInput->UpdateBooleanComponent(m_ulInputComponentHandle, true, -0.1);
pDriverInput->UpdateBooleanComponent(m_ulInputComponentHandle, false, 0);
```

**Comment by peroht** [comment]
> Any app using controller roles would need to call GetTrackedDeviceIndexForControllerRole and/or GetControllerRoleForTrackedDeviceIndex to do that. Those functions should already be obeying TrackedControllerRole_OptOut. Are you seeing a case where they aren't?

In my driver, I've seen that if I classify my custom device/controller as a `ETrackedDeviceClass::TrackedDeviceClass_Controller` it will lead to Unreal consuming the device and beliving it to be a normal vive controller, thus the hand gets lost and basically positioning/orienting it completely wrong. I do not seen this problem in games built with Unity. 

A couple of weeks ago I checkout out the controller filtering/assigning code in Unreal (4.18) and I yes it used GetControllerRoleForTrackedDeviceIndex(). I could however not see TrackedControllerRole_OptOut in there since this was added in openvr after 4.18 was released. I have yet to check the code for Unreal 4.19. 

If i change it to `ETrackedDeviceClass::TrackedDeviceClass_GenericTracker` however I no longer see these problems. Looking at the code comment:

`TrackedDeviceClass_GenericTracker = 3,		// Generic trackers, similar to controllers
`

Do you advice me to continue using `ETrackedDeviceClass::TrackedDeviceClass_GenericTracker` (although hypothetically my device is a kind of controller) instead of `ETrackedDeviceClass::TrackedDeviceClass_Controller`? 
Is there a limit to the amount of TrackedDeviceClass_Controller one should use? I guess we wont see people with three hands/controllers anytime soon though why a soft limit of 2 is sensible.. :)

Can we expect that the code comment above "similar to controllers" will mean that the TrackedDeviceClass_GenericTracker will have at least the same level of buttons as a TrackedDeviceClass_Controller?

> We added Prop_NeverTracked_Bool so that the status window would stop blinking annoyingly for gamepads. What support for that property would you expect in applications?
It is perfect for the status window for one of the drivers of our Omnideck, since it basically is a giant gamepad and should not need to have its pose updated/blink.

**Comment by bddckr** [comment]
@peroht Beware of some titles treating your `TrackedDeviceClass_GenericTracker` as Vive Trackers for body presence etc.

Limit of devices is only defined by
https://github.com/ValveSoftware/openvr/blob/d35c04ca3d7ddb762afa4a1c6c4a9d5de992cda7/headers/openvr.h#L156
AFAIK - the device's class shouldn't matter, though some titles use any device after two controllers as trackers for body presence or mixed reality features.

**Comment by peroht** [comment]
Yes, this is the dilemma I'm facing right now. Thus I tried a while with `TrackedDeviceClass_Controller` but that gives alot of problems with games written in Unreal (nothing against Unreal at all).
